### PR TITLE
Stefanozampini/snes

### PR DESCRIPTION
**Pull request by stefanozampini**

### CODE REVIEW

File: src/binding/petsc4py/src/petsc4py/PETSc/SNES.pyx [hunk 1]
```diff
@@ -252,58 +539,173 @@ cdef class SNES(Object):
...
+    def setInitialGuess(self, initialguess: SNESGuessFunction,
+                        args: tuple[Any, ...] | None = None,
+                        kargs: dict[str, Any] | None = None) -> None:
+        """Set the callback to compute the initial guess.
+
+        Logically collective.
+
+        Parameters
+        ----------
+        initialgues
```
Comment (mrogowski) [inline src/binding/petsc4py/src/petsc4py/PETSc/SNES.pyx:654]:
```suggestion
        initialness
```

Comment (mrogowski) [inline src/binding/petsc4py/src/petsc4py/PETSc/SNES.pyx:654]:
```suggestion
        initialguess
```

File: src/binding/petsc4py/src/petsc4py/PETSc/SNES.pyx [hunk 2]
```diff
@@ -144,57 +273,157 @@ cdef class SNES(Object):
...
+        """Set the number of levels to use with FAS.
+
+        Arguments
+        ---------
+        levels :
+          The number of levels
+        comms :
+          An optional sequence of communicators of length `levels`, or `None` for the default communicator `Sys.getDefaultComm`.
+
+        Collective.
```
Comment (mrogowski) [inline src/binding/petsc4py/src/petsc4py/PETSc/SNES.pyx:420]: should be above arguments

File: src/binding/petsc4py/src/petsc4py/PETSc/SNES.pyx [hunk 3]
```diff
@@ -346,10 +776,42 @@ cdef class SNES(Object):
             self.set_attr('__update__', None)
             CHKERR( SNESSetUpdate(self.snes, NULL) )
 
-    def getUpdate(self):
+    def getUpdate(self) -> SNESUpdateFunction:
+        """Return the callback to compute update at the beginning of the nonlinear step.
```
Comment (dham) [inline src/binding/petsc4py/src/petsc4py/PETSc/SNES.pyx:780]:
```suggestion
        """Return the callback to compute the update at the beginning of the nonlinear step.
```

File: src/binding/petsc4py/src/petsc4py/PETSc/SNES.pyx [hunk 4]
```diff
@@ -590,111 +1461,343 @@ cdef class SNES(Object):
...
+        --------
+        petsc.SNESSetForceIteration
+
+        """
         cdef PetscBool bval = asBool(force)
         CHKERR( SNESSetForceIteration(self.snes, bval) )
 
-    def setFunctionNorm(self, norm):
+    def setFunctionNorm(self, norm: float) -> None:
+        """Set the function norm.
```
Comment (dham) [inline src/binding/petsc4py/src/petsc4py/PETSc/SNES.pyx:1606]:
```suggestion
        """Set the function norm value. This is only of use to implementers of custom SNES types. 
```

File: src/binding/petsc4py/src/petsc4py/PETSc/SNES.pyx [hunk 5]
```diff
@@ -590,111 +1461,343 @@ cdef class SNES(Object):
...
+        setErrorIfNotConverged, petsc.SNESGetErrorIfNotConverged
+
+        """
         cdef PetscBool flag = PETSC_FALSE
         CHKERR( SNESGetErrorIfNotConverged(self.snes, &flag) )
         return toBool(flag)
 
-    def setIterationNumber(self, its):
+    def setIterationNumber(self, its: int) -> None:
+        """Set the current iteration number.
```
Comment (dham) [inline src/binding/petsc4py/src/petsc4py/PETSc/SNES.pyx:1566]:
```suggestion
        """Set the current iteration number. This is only of use to implementers of custom SNES types.
```

File: src/binding/petsc4py/src/petsc4py/PETSc/SNES.pyx [hunk 6]
```diff
@@ -590,111 +1461,343 @@ cdef class SNES(Object):
...
+
+        """
         cdef KSP ksp = KSP()
         CHKERR( SNESGetKSP(self.snes, &ksp.ksp) )
         PetscINCREF(ksp.obj)
         return ksp
 
-    def setUseEW(self, flag=True, *targs, **kargs):
+    def setUseEW(self, flag: bool = True, *targs: Any, **kargs: Any) -> None:
+        """Inform the solver to use the Eisenstat-Walker trick.
```
Comment (dham) [inline src/binding/petsc4py/src/petsc4py/PETSc/SNES.pyx:1733]:
```suggestion
        """Tell the solver to use the Eisenstat-Walker trick.
```

**Review by dham — DISMISSED** [review]

**Pull request merged**